internal/server/singleprocess/poll_application.go

```diff
@@ -67,34 +68,43 @@ func (a *applicationPoll) PollJob(
 	var jobList []*pb.QueueJobRequest
 
 	for _, app := range project.Applications {
-		job, err := a.buildPollJob(log, app)
+		jobs, err := a.buildPollJobs(log, app)
 		if err != nil {
 			return nil, err
 		}
 
-		jobList = append(jobList, job)
+		jobList = append(jobList, jobs...)
 	}
 
 	return jobList, nil
 }
 
-// buildPollJob will generate a job to queue a project on.
-// When determining what to generate a report on, either a Deployment or Release,
-// this job assumes that the _Release_ was the last operation taken on the application.
-// If there's a release, this job will queue a status report genreation on that.
-// Otherwise if there's just a deployment, it returns a job to generate a
-// status report on the deployment.
-func (a *applicationPoll) buildPollJob(
+// buildPollJobs will generate jobs to poll the latest deploy and release operations for a project.
+func (a *applicationPoll) buildPollJobs(
 	log hclog.Logger,
 	appl interface{},
-) (*pb.QueueJobRequest, error) {
+) ([]*pb.QueueJobRequest, error) {
 	app, ok := appl.(*pb.Application)
 	if !ok || app == nil {
 		log.Error("could not generate poll job for application, incorrect type passed in")
 		return nil, status.Error(codes.FailedPrecondition, "incorrect type passed into Application PollJob")
 	}
 	log = log.Named(app.Name)
 
+	// App polling needs the parent project to obtain its datasource
+	project, err := a.state.ProjectGet(&pb.Ref_Project{Project: app.Project.Project})
+	if err != nil {
+		return nil, err
+	}
+
+	// Application polling requires a remote data source, otherwise a status report
+	// cannot be generated without a project and its hcl context. This returns
+	// an error so we fail early instead of queueing an already broken job
+	if project.DataSource == nil {
+		log.Debug("cannot poll a job without a remote data source configured.")
+		return nil, status.Error(codes.FailedPrecondition, "application polling requires a remote data source")
```

Comment:
I wrote this originally but this should say `application status polling` to be more specific.

```suggestion
		return nil, status.Error(codes.FailedPrecondition, "application status polling requires a remote data source")
```